logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt

```diff
@@ -229,21 +232,40 @@ internal class MessageSenderImpl internal constructor(
             .flatMap { recipients ->
                 sessionEstablisher
                     .prepareRecipientsForNewOutgoingMessage(recipients)
-                    .map { failedToListUserIds ->
-                        recipients
-                    }
-                // TODO(federation) filter clients with failed to get prekeys and add persist in db
-            }.fold({
-                // TODO(federation) if (it is NetworkFailure.FederatedBackendError)
-                // TODO(federation) handle federated failure to filter clients and add to QualifiedMessageOption.IgnoreSome
-                Either.Left(it)
-            }, { recipients ->
+                    .flatMap { handleUsersWithNoClientsToDeliver(conversationId, message.id, it) }
+                    .map { recipients to getMessageOption(it, messageTarget) }
+            }.flatMap { (recipients, messageOption) ->
                 messageEnvelopeCreator
                     .createOutgoingEnvelope(recipients, message)
                     .flatMap { envelope ->
-                        trySendingProteusEnvelope(envelope, message, messageTarget)
+                        trySendingProteusEnvelope(envelope, message, messageOption, messageTarget)
                     }
-            })
+            }
+    }
+
+    private suspend fun handleUsersWithNoClientsToDeliver(
+        conversationId: ConversationId,
+        messageId: String,
+        usersWithoutSessions: UsersWithoutSessions
+    ): Either<CoreFailure, UsersWithoutSessions> = if (usersWithoutSessions.hasMissingSessions()) {
+        messageRepository.persistNoClientsToDeliverFailure(conversationId, messageId, usersWithoutSessions.users)
+            .flatMap { Either.Right(usersWithoutSessions) }
+    } else {
+        Either.Right(usersWithoutSessions)
+    }
+
+    private fun getMessageOption(
+        usersWithoutSessions: UsersWithoutSessions,
+        messageTarget: MessageTarget
+    ): MessageApi.QualifiedMessageOption = when (messageTarget) {
+        is MessageTarget.Client -> MessageApi.QualifiedMessageOption.IgnoreAll
```

Comment:
Should we silently not send a targeted message to specific client if we can't establish a session?